[NEW] stock_picking_wave_transfer_button], [FIX] stock_picking_wave_pack...

alfredoavanzosc opened the pull request:
...age_info

Por requerimientos de reclamacion CLM0237-Configuración de dimensiones de bultos en agrupación de albaranes (picking wave), se han quitado cosas del módulo stock_picking_wave_package_info, para metérselas al nuevo módulo stock_picking_wave_transfer_button.

Pedro... en el nuevo módulo, al heredar el wizard, la función default_get la he tenido que hacer con la api vieja, ya que cuando había puesto esta función con la api nueva, y llamaba con el método super a la función con la api vieja, en esta última función con api se modifica el contexto, y me decía que tururú.

También me gustaría comentarte otra cosa sobre el módulo nuevo, se ha metido un botón nuevo "Transferir", que lo que hace son las transferencias de todos los albaranes que hay definidos en la agrupación de albaranes.
Lo que me ha pedido Ana, es que si después de pinchar el botón "Transferir", todos los albaranes definidos en la agruación, quedan en estado "Transferido", quede automáticamente como "Realizado", la agrupación de albaranes.
El caso es que he hecho la prueba de programar al final del todo de la función del nuevo botón "Transferir", el recorrerme todos los albaranes de la agrupación, y el caso es que aunque se ha terminado todo el proceso, es decir, los albaranes ya estan como realizados, aunque accedo otra vez a la agrupación de los albaranes, estos no están con el nuevo estado, y por lo tanto no puedo cerrar la agrupacion. Joe... espero haberme explicado bien, ¿me puedes sugerir algo?.

Muchas gracias

Commit:
[NEW] stock_picking_wave_transfer_button], [FIX] stock_picking_wave_package_info

Commit:
[FIX] stock_picking_wave_transfer_button: travis error corrected

Commit:
[FIX] stock_picking_wave_transfer_button: travis error corredted

Commit:
[IMP] stock_picking_wave_transfer_button: has been added to context id of the wave.

Commit:
[IMP] stock_picking_wave_transfer_button: to create the new wave picking, the object is returned.

Commit:
[ADD] stock_picking_wave_parnter_carrier_filter, stock_picking_wave_button_partner_carrier_filter:
New modules to filter stock picking carrier wave by partner and carrier.

Commit:
[FIX] stock_picking_wave_button_partner_carrier_filter: Travis error corrected.

pedrobaeza commented on stock_picking_wave_button_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,20 @@
+installed:
+
+stock_picking_wave_transfer_button
+stock_picking_wave_partner_carrier_filter
```
`stock_picking_wave`

pedrobaeza replied:
Ah, vale, ya he visto que es otro módulo distinto (con nombre muy parecido), pero no veo por qué modularizar tanto. ¿No pueden ir juntos?

pedrobaeza commented on stock_picking_wave_button_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,20 @@
+Stock picking wave button partner carrier info
+==============================================
+
+This module is automatically installed when the following modules are
```
Pon esto en la sección "Installation"

alfredoavanzosc replied:
Donde esta la seccion "Installation"?, no veo ningún  ejemplo

pedrobaeza replied:
Mira, ésta es la plantilla de OCA que se utiliza, con las secciones recomendadas:

https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

pedrobaeza commented on stock_picking_wave_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,15 @@
+Stock picking wave partner carrier filter
```
Como el objetivo es dotar de la misma funcionalidad de los albaranes a las waves, mejor concentrar todos estos módulos en uno que se llame "stock_picking_wave_management".

pedrobaeza replied:
De todas formas, en concreto el carrier sí que debería ir por separado, ya que depende de delivery y además, habría que ver la conveniencia de ponerlo, ya que tú puedes mezclar más de un carrier/partner en un wave.

Por el momento, pon lo de transfer y demás en ese módulo que te comento.

alfredoavanzosc commented:
Pedro.... lo siento mucho, pero no entiendo. Esto es lo que tengo:

# stock_picking_wave_transfer_button

Este módulo permite comprobar disponibilidad, y transferir agrupaciones de albaranes.

# stock_picking_wave_package_info

Este módulo muestra los paquetes de todos los albaranes agrupados.

# stock_picking_wave_partner_carrier_filter

Este módulo permite en agrupaciones de albarnanes filtrar por empresa/transporte.

Hasta aquí estos 3 módulos son independientes, cada uno hace una cosa completamente distinta. Ahora el problema venía si se instalaban a la vez los módulos “stock_picking_wave_transfer_button”, y “stock_picking_wave_partner_carrier_filter”, por si se instalaban estos 2 módulos hice el módulo “stock_picking_wave_button_partner_carrier_info” que es autoinstalable.

No entiendo cuando me dices que los agrupe en el módulo “stock_picking_wave_management", porque de esta manera estás obligando a instalar “stock_picking_wave_trasfer_button”, y “stock_picking_wave_partner_carrier_filter”, cuando no tienen nada que ver uno con el otro.

Lo siento mucho.... pero no lo veo

pedrobaeza commented:
No, vale, es correcto lo que dices, pero como veo que ahora stock_picking_wave_package_info depende de stock_picking_wave_transfer_button, me había desconcertado un poco. ¿A qué se debe esa dependencia?

Por otro lado, ¿por qué necesitas el módulo puente?

Estas preguntas son por desconocimiento, eh!

alfredoavanzosc commented:
Ok, te explico. No es verdad que el módulo "stock_picking_wave_package_info" sea un módulo independiente, como tu bien has visto depende del módulo "stock_picking_wave_transfer_button". Este último módulo se puede decir que es el módulo base, el que siempre se va a instalar, ya que "stock_picking_wave" muestra los albaranes en el form sin más, y el módulo "stock_picing_wave_transfer_button", crea un notebook para meter estos albaranes en una pestaña.

pedrobaeza commented:
Vale, entonces mejor renombra _stock_picking_wave_transfer_button_ a _stock_picing_wave_management_, y explica en el README que ofrece una gestión de tareas que se harían desde albaranes en el propio wave.

Commit:
[IMP] stock_picking_wave_management: rename module before stock_picking_wave_transfer_button.
[IMP] stock_picking_wave_package_info: change dependency to stock_picking_wave_management.
[IMP] stock_picking_wave_button_partner_carrier_filter: change dependency to stock_picking_wave_management.

alfredoavanzosc commented:
Realizado todo lo solicitado por Pedro, y subidos los módulos.

Commit:
[IMP] stock_picking_wave_partner_carrier_filter: change domain to select the company

Commit:
[IMP] stock_picking_wave_partner_carrier_filter: change domain

anajuaristi commented:
@alfredoavanzosc casca travis. Echale un vistazo

Commit:
[FIX] stock_picking_wave_partner_carrier_filter: Travis error corredted.

pedrobaeza commented on stock_picking_wave_button_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,21 @@
+Stock picking wave button partner carrier info
+==============================================
+
+To create a new wave, data partner and carrier are updated from the origin
```
s/To create/When creating/

pedrobaeza commented on stock_picking_wave_button_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,21 @@
+Stock picking wave button partner carrier info
```
Este nombre no es correcto. Ya dijimos que el button iba a fuera, y la palabra filter no pinta nada. Debería ser _stock_picking_wave_partner_carrier_

pedrobaeza replied:
El nombre era asumiendo que era sólo un módulo. No entiendo por qué hay dos.

pedrobaeza commented on stock_picking_wave_button_partner_carrier_filter/README.rst:
```diff
@@ -0,0 +1,21 @@
+This module is automatically installed when these modules are installed:
+
+stock_picking_wave_management
+stock_picking_wave_partner_carrier_filter
```
Pero esto no dijimos que se podía quedar en un solo módulo?

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+
+    @api.one
+    def _count_confirmed_pickings(self):
+        self.num_confirmed = sum(1 for x in self.picking_ids if x.state ==
```
Puedes ponerlo mejor así:

```
self.num_confirmed = len(self.picking_ids.filtered(lambda x: x.state == 'confirmed'))
```

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+
+    @api.one
+    def _count_assigned_pickings(self):
+        self.num_assigned = sum(1 for x in self.picking_ids if x.state ==
```
Ídem

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+        compute="_count_assigned_pickings", string="Assigned pickings")
+
+    @api.one
+    def button_check_disponibility(self):
```
s/disponibility/availability

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,53 @@
+        pickings.action_assign()
+
+    def action_transfer(self, cr, uid, ids, context=None):
+        picking_obj = self.pool['stock.picking']
```
Pon un comentario para comentar que es API vieja porque perdía el context.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+        for picking in self.picking_ids:
+            processed_ids = []
+            # Create new and update existing pack operations
+            for lstits in [self.item_ids, self.packop_ids]:
```
Si es el mismo recordset, simplemente une ambos y lo puedes recorrer todo junto:

```
for prod in self.item_ids + self.packop_ids:
```

alfredoavanzosc replied:
Pedro... esto esta sacado de Odoo, y no es el mismo recordset.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+            # Create new and update existing pack operations
+            for lstits in [self.item_ids, self.packop_ids]:
+                for prod in lstits:
+                    if prod.picking_id.id == picking.id:
```
Y puesto que aquí vas buscando un elemento en concreto, haz un filtered y ya está:

```
op_list = self.item_ids + self.packop_ids
prod = op_list.filtered(lambda x: x.picking_id.id == picking.id)
if prod:
    ...
```

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+                packop.unlink()
+            # Execute the transfer of the picking
+            picking.do_transfer()
+            new_pickings = self._find_new_pickings(new_pickings, picking)
```
Esto hace que sólo se coja el último valor de new_pickings, por lo que no vale para más de un registro. Aunque es un caso que no creo que se dé, deberías indentar lo de "if new_picking..."

alfredoavanzosc replied:
No Pedro, no se queda con el último valor, a la función _find_new_pickings se le pasa el campo "new_pickings", y también recibe este mismo parámetro, pero antes se le ha hecho un ".extend", por lo que no se queda solo con el último.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+        picking_obj = self.env['stock.picking']
+        cond = [('backorder_id', '=', picking.id)]
+        pickings = picking_obj.search(cond)
+        if not isinstance(pickings, list):
```
Esto no va a ocurrir, ya que con API nueva siempre devuelve un recordset.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+        pickings = picking_obj.search(cond)
+        if not isinstance(pickings, list):
+            pickings = [pickings]
+        new_pickings.extend([x.id for x in pickings])
```
No mezcles recordset con ids. Utiliza siempre recordset, y así te ahorras tener que hacer posibles browse en otro sitio. Con nueva API, no te interesa mantener listas de IDs, si no recordsets.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,138 @@
+                'state': 'draft'
+                }
+        new_wave = wave_obj.create(vals)
+        pickings = picking_obj.browse(new_pickings)
```
Ves, aquí estás teniendo que hacer un browse. Además, la nomenclatura no es correcta, ya que debería haber sido _new_picking_ids_, pero lo dicho, nunca más ids.

pedrobaeza commented:
Alfredo, acaban de publicar en OCA un módulo que hace lo mismo. Echa por favor un vistazo a ver si cubre lo que hacemos con el management, y así no duplicamos esfuerzos:

https://github.com/OCA/stock-logistics-workflow/pull/87

CC @anajuaristi

anajuaristi commented:
@pedrobaeza tendría que verlo instalado pero en principio, no hace lo mismo
Lo que hace es "empaquetar" todos los albaranes y meterlos en un pallet.
El nuestro es más genérico. Quizás puedan ser complementarios pero en principio el nuestro vale solo para seleccionar rápidamente qué albaranes quieres meter en la agrupación y unificar los botones de albaranes como comprobar disponibilidad y transferir. 

Si quieres lo ponemos en nuestro server, lo pruebo y te digo más exactamente las diferencias.

Commit:
[IMP] stock_picking_wave_management: Added to filter by partner.
[IMP] stock_picking_wave_carrier_filter: Only to filter by carrier.
[DEL] stock_picking_wave_button_partner_carrier_filter: This module disappears.

alfredoavanzosc commented:
Realizado lo solicitado por Pedro, y subidos los módulos.

Commit:
[FIX] stock_picking_wave_management: bug corrected.

pedrobaeza commented on stock_picking_wave_carrier_filter/README.rst:
```diff
@@ -0,0 +1,13 @@
+Stock picking wave carrier filter
```
El nombre debería ser _stock_picking_wave_delivery_ para seguir con la nomenclatura del módulo principal delivery.

pedrobaeza commented on stock_picking_wave_carrier_filter/models/stock_picking_wave.py:
```diff
@@ -0,0 +1,23 @@
+    def onchange_carrier(self):
+        self.ensure_one()
+        result = self.onchange_partner()
+        if 'domain' in result and self.carrier:
```
Si no existe el dominio, habría que añadir la clave entonces.

pedrobaeza commented on stock_picking_wave_management/__openerp__.py:
```diff
@@ -0,0 +1,35 @@
+    'version': "1.0",
+    'author': 'OdooMRP team',
+    'website': "http://www.odoomrp.com",
+    'category': 'Stock Management',
```
Sería Warehouse Management

pedrobaeza commented on stock_picking_wave_carrier_filter/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+    'version': "1.0",
+    'author': 'OdooMRP team',
+    'website': "http://www.odoomrp.com",
+    'category': 'Stock Management',
```
Sería Warehouse Management.

pedrobaeza commented on stock_picking_wave_management/i18n/stock_picking_wave_management.pot:
```diff
@@ -0,0 +1,119 @@
+
+#. module: stock_picking_wave_management
+#: view:stock.picking.wave:stock_picking_wave_management.view_picking_wave_form_inh_management
+msgid "Check Disponibility"
```
Cambiar

pedrobaeza commented on stock_picking_wave_management/i18n/es.po:
```diff
@@ -0,0 +1,119 @@
+
+#. module: stock_picking_wave_management
+#: view:stock.picking.wave:stock_picking_wave_management.view_picking_wave_form_inh_management
+msgid "Check Disponibility"
```
Cambiar

pedrobaeza commented on stock_picking_wave_management/models/__init__.py:
```diff
@@ -0,0 +1,6 @@
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+from . import stock
```
El archivo debería ser stock_picking_wave y separar la otra clase en stock_move. ¿Por qué seguir uniendo esto en archivos gigantes con nombre genérico?

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,69 @@
+    @api.one
+    def button_check_availability(self):
+        picking_obj = self.env['stock.picking']
+        picking_ids = [picking.id for picking in
```
Por qué sacar los ids para luego volver a hacer browse. ¿Es que no recuerdas el uso de filtered?
`pickings = self.picking_ids.filtered(lambda x: x.state == 'confirmed')`

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,69 @@
+    def action_transfer(self, cr, uid, ids, context=None):
+        picking_obj = self.pool['stock.picking']
+        wave = self.browse(cr, uid, ids[0], context=context)
+        picking_ids = [picking.id for picking in
```
Ídem

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,69 @@
+        self.ensure_one()
+        cond = [('state', 'not in', ('done', 'cancel'))]
+        if self.partner:
+            if self.partner.child_ids:
```
No necesitas la doble condición. Esto ya lo hemos dicho muchas veces.

pedrobaeza commented on stock_picking_wave_management/models/stock.py:
```diff
@@ -0,0 +1,69 @@
+        cond = [('state', 'not in', ('done', 'cancel'))]
+        if self.partner:
+            if self.partner.child_ids:
+                ids = map(lambda x: x['id'], self.partner.child_ids)
```
¡¡¿¿Por qué esto??!! Con lo fácil que es `self.partner.child_ids.ids`...

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,139 @@
+        item_ids = []
+        packop_ids = []
+        picking_ids = context.get('active_ids', [])
+        for picking in picking_ids:
```
Pon la variable como `picking_id` para no perder la referencia a que es un id.

pedrobaeza commented on stock_picking_wave_management/wizard/stock_transfer_details.py:
```diff
@@ -0,0 +1,139 @@
+                cr, uid, fields, context=c)
+            lines = res.get('item_ids')
+            for line in lines:
+                line.update({'picking_id': picking})
```
¿Pero por qué sigues sin utilizar asignaciones directas? `line['picking_id'] = picking`

Commit:
[IMP] stock_picking_wave_delivery: module renamed.

alfredoavanzosc commented:
Pedro.. yo me tengo que ir, haz por favor el merge para que Ana mañana lo pueda enseñar, y mañana a primera hora, corrigo todo lo que me has pasado ahora a última hora.

pedrobaeza commented:
Se pone, pero sin hacer merge. Que se ponga en other-addons. Luego se borran esos módulos y que coja los de odoomrp-wip.

Commit:
[FIX] stock_picking_wave_management: Improve programming.

Commit:
[FIX] stock_picking_wave_management: travis correct errors.

alfredoavanzosc commented:
Realizado lo solicitado por Pedro, subidos los módulos.

pedrobaeza commented:
:+1: cuando Travis dé verde
